### ISSUE TITLE
Doesn't respect pipe() groups or `group:` tags & --dry-run output differs from real-world group execution

### ISSUE THREAD

**Issue by semenko** [snakemake/snakemake#1822]
**Snakemake version**
7.12.1

**Describe the bug**
I have a Snakemake pipeline where different rules with `temp(pipe())` linking them are not run in the same group.

Perplexingly, when I run snakemake with `--dry-run`, the rules are (I think always?) shown correctly grouped, but when I run the job without the dry-run flag, the rules are **not** consistently correctly grouped -- and one `pipe()` job and its consumer aren't run at the same time (so the execution stalls).

I specifically see this when running with ```snakemake --cores 40 --use-conda --rerun-incomplete -p --rerun-triggers mtime```

**Has anyone seen this issue?** Perhaps this is related to #1813?

------------------

In this example, the rule **mark_nonconverted** should pipe into **samtools_fixmate** -- but sometimes these jobs aren't run in the same group. I tried explicitly setting `group: something` to each rule, but that didn't work (and the jobs were sometimes still separated).

**Rules**

```
rule mark_nonconverted:
    conda: "envs/env.yaml"
    input:
        aligned_sam = "{DATA_PATH}/{experiment}/{sample}/bwameth/raw.sam",
        reference_genome = rules.mask_reference_fasta.output.masked_reference_genome
    output:
        temp(pipe("{DATA_PATH}/{experiment}/{sample}/bwameth/nonconv-marked.sam"))
    log:
        nonconverted = "{DATA_PATH}/{experiment}/{sample}/bwameth/nonconverted.log.txt"
    params:
        threshold = 3 # If 3 nonconverted Cs on a read, consider it nonconverted
    shell:
        "cat {input.aligned_sam} | mark-nonconverted-reads.py --c_count {params.threshold} --reference {input.reference_genome} 2>{log.nonconverted} >{output}"

rule samtools_fixmate:
    conda: "envs/env.yaml"
    input:
        "{DATA_PATH}/{experiment}/{sample}/bwameth/nonconv-marked.sam"
    output:
        temp(pipe("{DATA_PATH}/{experiment}/{sample}/samtools/fixmate.sam"))
    log:
        fixmate = "{DATA_PATH}/{experiment}/{sample}/samtools/fixmate.log.txt"
    shell:
        "cat {input} | samtools fixmate -u -m --output-fmt SAM --threads {threads} - - 2>{log.fixmate} >{output}"
```

**Logs**
One run where it didn't work -- the rule `samtools_fixmate` wasn't spawned to consume `mark_nonconverted`
```
group job f42a64cf-dfc7-4e2f-b5bc-16157a8689bc (jobs in lexicogr. order):

    [Mon Aug 22 12:37:13 2022]
    rule bwa_meth:
        input: ../data/test/Test_SAMPLE1/fastp/interleaved.fa, ../data/reference/GRCh38-DAC-U2AF1.fna, ../data/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t, ../data/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.amb, ../data/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.ann, ../data/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.bwt, ../data/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.pac, ../data/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.sa
        output: ../data/test/Test_SAMPLE1/bwameth/raw.sam (pipe)
        log: ../data/test/Test_SAMPLE1/bwameth/log.txt
        jobid: 7
        reason: Missing output files: ../data/test/Test_SAMPLE1/bwameth/raw.sam; Input files updated by another job: ../data/test/Test_SAMPLE1/fastp/interleaved.fa
        wildcards: DATA_PATH=../data, experiment=test, sample=Test_SAMPLE1
        threads: 30
        resources: tmpdir=/tmp

    bwameth.py -p -t 30 --read-group '@RG\tID:Test_SAMPLE1\tSM:Test_SAMPLE1' --reference ../data/reference/GRCh38-DAC-U2AF1.fna ../data/test/Test_SAMPLE1/fastp/interleaved.fa 2>../data/test/Test_SAMPLE1/bwameth/log.txt >../data/test/Test_SAMPLE1/bwameth/raw.sam

    [Mon Aug 22 12:37:13 2022]
    rule fastp:
        input: ../data/test/Test_SAMPLE1/raw/validated.md5, ../data/test/Test_SAMPLE1/raw/1.fastq.gz, ../data/test/Test_SAMPLE1/raw/2.fastq.gz
        output: ../data/test/Test_SAMPLE1/fastp/interleaved.fa (pipe), ../data/test/Test_SAMPLE1/fastp/failed.fa, ../data/test/Test_SAMPLE1/fastp/fastp.json, ../data/test/Test_SAMPLE1/fastp/fastp.html
        log: ../data/test/Test_SAMPLE1/fastp/fastp.log.txt
        jobid: 8
        reason: Missing output files: ../data/test/Test_SAMPLE1/fastp/interleaved.fa; Input files updated by another job: ../data/test/Test_SAMPLE1/raw/validated.md5
        wildcards: DATA_PATH=../data, experiment=test, sample=Test_SAMPLE1
        resources: tmpdir=/tmp


        fastp --in1 ../data/test/Test_SAMPLE1/raw/1.fastq.gz --in2 ../data/test/Test_SAMPLE1/raw/2.fastq.gz         --trim_front1 10 --trim_tail1 10         --trim_front2 10 --trim_tail2 10         --length_required 10 --disable_quality_filtering --trim_poly_g         --adapter_sequence "AGATCGGAAGAGCACACGTCTGAACTCCAGTCA" --adapter_sequence_r2 "AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT"         --json "../data/test/Test_SAMPLE1/fastp/fastp.json" --html "../data/test/Test_SAMPLE1/fastp/fastp.html"         --thread 1 --verbose --failed_out ../data/test/Test_SAMPLE1/fastp/failed.fa --stdout         2>../data/test/Test_SAMPLE1/fastp/fastp.log.txt >../data/test/Test_SAMPLE1/fastp/interleaved.fa


    [Mon Aug 22 12:37:13 2022]
    rule mark_nonconverted:
        input: ../data/test/Test_SAMPLE1/bwameth/raw.sam, ../data/reference/GRCh38-DAC-U2AF1.fna
        output: ../data/test/Test_SAMPLE1/bwameth/nonconv-marked.sam (pipe)
        log: ../data/test/Test_SAMPLE1/bwameth/nonconverted.log.txt
        jobid: 6
        reason: Missing output files: ../data/test/Test_SAMPLE1/bwameth/nonconv-marked.sam; Input files updated by another job: ../data/test/Test_SAMPLE1/bwameth/raw.sam
        wildcards: DATA_PATH=../data, experiment=test, sample=Test_SAMPLE1
        resources: tmpdir=/tmp

    cat ../data/test/Test_SAMPLE1/bwameth/raw.sam | /home/nsemenkovich/sandbox/mark-nonconverted-reads/mark-nonconverted-reads.py --c_count 3 --reference ../data/reference/GRCh38-DAC-U2AF1.fna 2>../data/test/Test_SAMPLE1/bwameth/nonconverted.log.txt >../data/test/Test_SAMPLE1/bwameth/nonconv-marked.sam
Activating conda environment: .snakemake/conda/419b065dcbfe4689fc19ca93d59b67c1
Activating conda environment: .snakemake/conda/419b065dcbfe4689fc19ca93d59b67c1
Activating conda environment: .snakemake/conda/419b065dcbfe4689fc19ca93d59b67c1
…
[This just stalls -- since the nonconv-marked.sam pipe is never consumed.
```

On a subsequent execution (not shown), `samtools_fixmate` was spawned, and the pipeline ran fine. There were no changes to the Snakefile in the meantime, I just ran it twice.

**Comment by semenko** [comment]
Ok, it also groups pipe jobs incorrectly sometimes with `--dry-run` too -- very perplexing. Confirmed on the latest version **7.12.1**

Here's an example output of `snakemake --cores 40 --use-conda --rerun-incomplete -p --rerun-triggers mtime --dry-run`

```

[Mon Aug 22 14:33:47 2022]

group job 94f2c0ad-e54f-4f4c-aa06-b88b35e11af8 (jobs in lexicogr. order):

…

    [Mon Aug 22 14:33:47 2022]
    rule mark_nonconverted:
        input: ../data/test/Test_SAMPLE1/samtools/raw.sam, ../data/reference/GRCh38-DAC-U2AF1.fna
        output: ../data/test/Test_SAMPLE1/samtools/nonconv-marked.sam (pipe)
        log: ../data/test/Test_SAMPLE1/logs/mark-nonconverted.log.txt
        jobid: 6
        reason: Missing output files: ../data/test/Test_SAMPLE1/samtools/nonconv-marked.sam; Input files updated by another job: ../data/test/Test_SAMPLE1/samtools/raw.sam
        wildcards: DATA_PATH=../data, experiment=test, sample=Test_SAMPLE1
        resources: tmpdir=/tmp

    cat ../data/test/Test_SAMPLE1/samtools/raw.sam | /home/nsemenkovich/sandbox/mark-nonconverted-reads/mark-nonconverted-reads.py --c_count 3 --reference ../data/reference/GRCh38-DAC-U2AF1.fna 2>../data/test/Test_SAMPLE1/logs/mark-nonconverted.log.txt >../data/test/Test_SAMPLE1/samtools/nonconv-marked.sam
Would remove temporary output ../data/test/Test_SAMPLE1/samtools/raw.sam
[Mon Aug 22 14:33:47 2022]

group job 4562317e-8248-48cc-b632-78c76b96f8f5 (jobs in lexicogr. order):

    [Mon Aug 22 14:33:47 2022]
    rule samtools_fixmate:
        input: ../data/test/Test_SAMPLE1/samtools/nonconv-marked.sam
        output: ../data/test/Test_SAMPLE1/samtools/fixmate.sam (pipe)
        log: ../data/test/Test_SAMPLE1/logs/samtools-fixmate.log.txt
        jobid: 5
        reason: Missing output files: ../data/test/Test_SAMPLE1/samtools/fixmate.sam; Input files updated by another job: ../data/test/Test_SAMPLE1/samtools/nonconv-marked.sam
        wildcards: DATA_PATH=../data, experiment=test, sample=Test_SAMPLE1
        resources: tmpdir=/tmp


…

```

I'm not sure how it's possible that these two jobs are in different groups, since the pipe() output of one is the input of the next.

**Comment by semenko** [comment]
My suspicion is this non-deterministic behavior has something to do with **temp(pipe())** -- snakemake suggests different temporary files to remove each --dry-run execution:
 
```
$ snakemake --cores 40 --use-conda --rerun-incomplete -p --rerun-triggers mtime --dry-run 2>&1 | grep "Would remove"
Would remove temporary output ../data/test/Test_SAMPLE1/samtools/pos-sorted.sam
$ snakemake --cores 40 --use-conda --rerun-incomplete -p --rerun-triggers mtime --dry-run 2>&1 | grep "Would remove"
Would remove temporary output ../data/test/Test_SAMPLE1/samtools/fixmate.sam
Would remove temporary output ../data/test/Test_SAMPLE1/samtools/pos-sorted.sam
$ snakemake --cores 40 --use-conda --rerun-incomplete -p --rerun-triggers mtime --dry-run 2>&1 | grep "Would remove"
Would remove temporary output ../data/test/Test_SAMPLE1/samtools/raw.sam
Would remove temporary output ../data/test/Test_SAMPLE1/samtools/pos-sorted.sam
```

**Comment by fgvieira** [comment]
Did you try changing all `temp(pipe())` to just `pipe()`?

**Comment by semenko** [comment]
Yes, that didn't resolve the issue unfortunately, and I simply abandoned pipe() jobs and put everything into one shell command.

**Comment by fgvieira** [comment]
Can you try with snakemake version `7.3.2`? I might have a similar problem that appeared first on version `7.3.3`.

**Comment by semenko** [comment]
Closed by #1846 ?
